### PR TITLE
AutoYaST Upgrade profile

### PR DESCRIPTION
**Pull request by rhafer**
This is the contains the base work for an autoyast profile to upgrade a SLE11 openstack controller to SLE12-SP1.

Currently this just creates autoyast-upgrade.xml just in parallel to the "normal" autoyast profile and hardcodes the autoupgrade parameter into the pxe config for all nodes that get allocated. This was just done to be able to test the profile. The final solution will have to handle it differently of course. This PR is just meant to be input for whoever is going to implement the "upgrade" state in crowbar.

### CODE REVIEW

File: chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+      <id>none</id>
+    </mouse>
+    <proposals config:type="list"/>
+    <signature-handling>
+      <accept_file_without_checksum config:type="boolean">true</accept_file_without_checksum>
+      <accept_non_trusted_gpg_key config:type="boolean">true</accept_non_trusted_gpg_key>
+      <accept_unknown_gpg_key config:type="boolean">true</accept_unknown_gpg_key>
+      <accept_unsigned_file config:type="boolean">true</accept_unsigned_file>
+      <accept_verification_failed config:type="boolean">true</accept_verification_failed>
+      <import_gpg_key config:type="boolean">true</import_gpg_key>
```
Comment (dirkmueller) [inline chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb]: do we really need those settings to be enabled?

Comment (rhafer) [inline chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb]: Actually I don't really know. Those are just copied from the non-upgrade template. If we don't need them there anymore, we don't need them here.

File: chef/cookbooks/provisioner/recipes/update_nodes.rb [hunk 2]
```diff
@@ -257,7 +257,7 @@
           end
 
         when os =~ /^(open)?suse/
-          append << "install=#{install_url} autoyast=#{node_url}/autoyast.xml"
+          append << "install=#{install_url} autoyast=#{node_url}/autoyast-upgrade.xml autoupgrade=1"
```
Comment (mjura) [inline chef/cookbooks/provisioner/recipes/update_nodes.rb]: Shouldn't we check somehow if this upgrade is needed otherwise we could use standard `autoyast.xml` template

Comment (rhafer) [inline chef/cookbooks/provisioner/recipes/update_nodes.rb]: Yes of course. But the underlying mechanism required for that in crowbar are still being worked out (e.g. here: https://github.com/crowbar/barclamp-updater/pull/11) and implemented yet. This commit here contains just the bare minimum that we needed to create/test an AutoYaST profile for upgrading. It's not meant to be merged in this form (see my initial comment)

File: chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        <product><%= name %></product>
+        <product_dir>/</product_dir>
+        <media_url><%= @repos[name][:url] %></media_url>
+        <!-- See bnc#787157 for discussion on ask_on_error -->
+        <ask_on_error config:type="boolean"><%= @repos[name][:ask_on_error] || 'false' %></ask_on_error>
+      </listentry>
+    <% end %>
+    </add_on_products>
+  </add-on>
+  <timezone>
```
Comment (mjura) [inline chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb]: I think this is not needed, if we want to do only upgrade.

File: chef/cookbooks/provisioner/recipes/update_nodes.rb (hunk 2 above)
Comment (mjura) [inline chef/cookbooks/provisioner/recipes/update_nodes.rb]: Ok, ack

**Comment by mjura** [comment]
LGTM

File: chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    <patterns config:type="list">
+      <pattern>Minimal</pattern>
+      <pattern>base</pattern>
+    </patterns>
+    <!-- Avoid default Cloud pattern "cloud_admin" to be installed on all nodes -->
+    <remove-patterns config:type="list">
+	<pattern>cloud_admin</pattern>
+    </remove-patterns>
+    <!-- Cleanup leftover Cloud 5 packages -->
+    <remove-packages config:type="list">
```
Comment (vuntz) [inline chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb]: Have you generated this list from a node where you had installed every package from Cloud 5?

File: chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+<% end -%>
+    </packages>
+    <patterns config:type="list">
+      <pattern>Minimal</pattern>
+      <pattern>base</pattern>
+    </patterns>
+    <!-- Avoid default Cloud pattern "cloud_admin" to be installed on all nodes -->
+    <remove-patterns config:type="list">
+	<pattern>cloud_admin</pattern>
+    </remove-patterns>
```
Comment (vuntz) [inline chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb]: Why is this needed? Is this just to be safe?

File: chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb (hunk 4 above)
Comment (rhafer) [inline chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb]: The Cloud5 node was basically an  HA controller node, that had all barclamps deployed. We didn't install everything that's on the media though. The list was basically gathered by doing this search on the upgraded node:

```
zypper se -t package -s |grep \(System | cut -d \| -f 2
```


But looking at the package diff between a node upgraded from Cloud 5 and one cleanly installed from the latested Cloud6 image there seems to be a lot more potential for cleanup (maybe even before the upgrade is done).

**Comment by rhafer** [comment]
I rebased this on top of #177 so please ignore those changes for now.

File: chef/cookbooks/provisioner/recipes/update_nodes.rb [hunk 6]
```diff
@@ -282,9 +283,11 @@
             cloud_available = true if name.include? "Cloud"
           end
 
+          autoyast_template="autoyast"
+          autoyast_template="autoyast-upgrade" if mnode[:state] == "os_upgrading"
```
Comment (jsuchome) [inline chef/cookbooks/provisioner/recipes/update_nodes.rb]: Place to use ternary operator?

File: crowbar_framework/config/locales/crowbar/en.yml [hunk 7]
```diff
@@ -331,6 +331,7 @@ en:
         unknown: '%{count} Unknown'
         unready: '%{count} Not Ready'
         pending: '%{count} Pending'
+        crowbar_upgrade: '%{count} Upgrade'
```
Comment (mmnelemane) [inline crowbar_framework/config/locales/crowbar/en.yml]: Question: Wont you need to add "Upgrade" in # Global state list too in this file ? (around Line 627)

**Comment by tboerger** [comment]
Some offenses can be fixed but :+1:

File: chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb (hunk 5 above)
Comment (rhafer) [inline chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb]: IIRC YaST  was selecting the default pattern for add-on when upgrading. But it seems we removed the default pattern options from the image meanwhile so I guess this can go away now.

**Comment by rhafer** [comment]
> Some offenses can be fixed

Yeah, will try to address some of that. But hound is now complaining about lines > 80 that seems really wrong and is not what we configured.

File: chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb (hunk 3 above)
Comment (rhafer) [inline chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb]: Yeah, I was unsure about this one. But it's an allowed autoyast section for the upgrade and doesn't really hurt. I can remove it though if you really insist :)

File: chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb (hunk 5 above)
Comment (rhafer) [inline chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb]: @vuntz this one is now fixed

**Comment by mjura** [comment]
Except that hound is bitching +1

File: crowbar_framework/app/models/crowbar_service.rb [hunk 8]
```diff
@@ -176,6 +176,48 @@ def create_proposal
...
+
+    # wait for the pxe_config to be updated, then reboot the nodes
+    discovery_dir = "#{NodeObject.admin_node[:provisioner][:root]}/discovery/"
+    pxecfg_subdir = "bios/pxelinux.cfg"
+
+    all_nodes.each do |node|
+      next if node.admin? || node["platform"] == "windows"
+
+      boot_ip_hex = node["crowbar"]["boot_ip_hex"]
+      node_arch = node["kernel"]["machine"] rescue "x86_64"
```
Comment (rhafer) [inline crowbar_framework/app/models/crowbar_service.rb]: @tboerger Do you think this needs to be addressed. If yes, what's a good way of fixing it?

Comment (mmnelemane) [inline crowbar_framework/app/models/crowbar_service.rb]: @tboerger, @ralf: How about
    node_arch = node["kernel"]["machine"] || "x86_64"  ? Its similar to what is suggested in the link by hound. "No exception for control flow."

**Comment by mmnelemane** [comment]
+1

File: crowbar_framework/app/models/crowbar_service.rb (hunk 8 above)
Comment (rhafer) [inline crowbar_framework/app/models/crowbar_service.rb]: That still throws an exception when node["kernel"] in nil

Comment (vuntz) [inline crowbar_framework/app/models/crowbar_service.rb]: `node.fetch("kernel", {}).fetch("machine", "x86_64")`

Comment (vuntz) [inline crowbar_framework/app/models/crowbar_service.rb]: That being said, why would node["kernel"] be nil?

Comment (rhafer) [inline crowbar_framework/app/models/crowbar_service.rb]: @vuntz Good question, but why would node["kernel"]["machine"] else throw an excpetion? 

But this check might really be unneeded here. I admittedly copied it from somewhere else (provisioner cookbook)

**Comment by rhafer** [comment]
Note, this is still not bring the nodes into the "ready" state. For that I need to deactivate all the non-core barclamps before (or as part of) calling ```prepare_nodes_for_os_upgrade```, but I'd prefer to create another PR for that.

File: chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb (hunk 3 above)
Comment (jsuchome) [inline chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb]: I think that when we are upgrading, it's really best to minimize the profile to stuff that needs changing. As this should stay as it is, I'm also for removing.

File: crowbar_framework/app/models/crowbar_service.rb [hunk 9]
```diff
@@ -176,6 +176,47 @@ def create_proposal
...
+      node_arch = node["kernel"]["machine"]
+      pxe_conf = "#{discovery_dir}/#{node_arch}/#{pxecfg_subdir}/#{boot_ip_hex}"
+      ready_for_reboot = false
+
+      while Time.now.to_i < Time.now.to_i + 120 && !ready_for_reboot
+        @logger.debug("waiting for pxe configuration to be updated for #{node.name}")
+        if File.file?(pxe_conf)
+          File.open(pxe_conf).each_line do |line|
+            line.chomp!
+            @logger.debug("line: #{line}")
```
Comment (jsuchome) [inline crowbar_framework/app/models/crowbar_service.rb]: Forgotten debug line? (Or maybe you want to have it there...)

File: crowbar_framework/app/models/crowbar_service.rb [hunk 10]
```diff
@@ -176,6 +176,47 @@ def create_proposal
...
+            if line =~ /^DEFAULT\s+.+_install$/
```
Comment (jsuchome) [inline crowbar_framework/app/models/crowbar_service.rb]: Wouldn't it be possible to grep the file instead?

**Comment by jsuchome** [comment]
I do not see any problems here (but I haven't test it)

File: crowbar_framework/app/models/crowbar_service.rb (hunk 9 above)
Comment (rhafer) [inline crowbar_framework/app/models/crowbar_service.rb]: No, that should not be there. I'll update.

File: crowbar_framework/app/models/crowbar_service.rb (hunk 10 above)
Comment (rhafer) [inline crowbar_framework/app/models/crowbar_service.rb]: grep, like calling out the external tool? I think that's really ugly.

**Comment by rhafer** [comment]
Addressed the last comments and squashed the first two commits into a single one.

**Comment by jsuchome** [comment]
+1 (without testing)

File: chef/cookbooks/provisioner/recipes/update_nodes.rb [hunk 11]
```diff
@@ -258,7 +258,8 @@
 
         when os =~ /^(open)?suse/
           append << "install=#{install_url} autoyast=#{node_url}/autoyast.xml"
-          append << "ifcfg=dhcp4 netwait=60"
+          append << " ifcfg=dhcp4 netwait=60"
+          append << " autoupgrade=1" if mnode[:state] == "os_upgrading"
```
Comment (vuntz) [inline chef/cookbooks/provisioner/recipes/update_nodes.rb]: I don't think you need the leading ` `: we are joining the items of `append` (`append.join(" ")`).

**Comment by vuntz** [comment]
If possible, I'd like to not use `_` but `-` in the state names. We allowed that to move forward for `crowbar_upgrade` (due to the led / pie), but I don't see anything depending on this for these states.

File: crowbar_framework/app/models/crowbar_service.rb [hunk 12]
```diff
@@ -176,6 +176,46 @@ def create_proposal
     base
   end
 
+  def prepare_nodes_for_os_upgrade
+    all_nodes = NodeObject.all
```
Comment (vuntz) [inline crowbar_framework/app/models/crowbar_service.rb]: What about:
```
upgrade_nodes = NodeObject.all.reject { |node| node.admin? || node[:platform] == "windows" }
```

File: crowbar_framework/app/models/crowbar_service.rb [hunk 13]
```diff
@@ -176,6 +176,46 @@ def create_proposal
...
+          end.close
+        end
+        sleep(5) unless ready_for_reboot
+      end
+      if ready_for_reboot
+        @logger.debug("Rebooting node #{node.name} for operating system upgrade")
+        node.ssh_cmd("/sbin/reboot")
+      end
+    end
+    true
```
Comment (vuntz) [inline crowbar_framework/app/models/crowbar_service.rb]: Should we return `false` if one node has not rebooted?

File: chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+
+            if ! curl -s -o /mnt/root/.ssh/authorized_keys.wget \
+                http://$HTTP_SERVER/authorized_keys ||\
+                grep -q "Error 404" /mnt/root/.ssh/authorized_keys.wget; then
+                rm -f /mnt/root/.ssh/authorized_keys.wget
+            else
+                chmod 644 /mnt/root/.ssh/authorized_keys
+                cat /mnt/root/.ssh/authorized_keys.wget >> /mnt/root/.ssh/authorized_keys
+                rm -f /mnt/root/.ssh/authorized_keys.wget
+            fi
```
Comment (vuntz) [inline chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb]: Doesn't really matter, but more out of curiosity: this bit (for authorized keys) should not be needed, since the old file is still there, right?

**Comment by vuntz** [comment]
For the record, I can live with the PR being merged and my comments addressed in a followup PR.

File: chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb (hunk 3 above)
Comment (rhafer) [inline chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb]: Ok, will remove that.

File: chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb (hunk 14 above)
Comment (rhafer) [inline chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb]: Right. I think the same is true for autoyast_set_hostentries.sh, I'll remove those two here.

File: crowbar_framework/app/models/crowbar_service.rb (hunk 13 above)
Comment (rhafer) [inline crowbar_framework/app/models/crowbar_service.rb]: Probably, but currently I have the problem that the ```node.ssh_cmd("/sbin/reboot")``` call on SLE12 systems will always return an error. I guess this is because it cause the ssh connection to be killed. Might need to figure something out for that.

**Comment by rhafer** [comment]
> If possible, I'd like to not use _ but - in the state names. We allowed that to move forward for crowbar_upgrade (due to the led / pie), but I don't see anything depending on this for these states.

Ok, I'll switch that.
